File: Data Structures & Algorithms/rotate-matrix/submission-0.cs
public class Solution {
    public void Rotate(int[][] matrix) {
        Array.Reverse(matrix);
        for(int i=0; i<matrix.Length; i++) {
            for(int j = i; j < matrix[0].Length; j++) {
                int temp = matrix[j][i];
                matrix[j][i] = matrix[i][j];
                matrix[i][j] = temp;
            }
        }
    }
}

// 1 2 3
// 4 5 6
// 7 8 9 

// 7 4 1
// 8 5 2
// 9 6 3

// 7 8 9
// 4 5 6
// 1 2 3

// 7 4 1
// 8 5 2 
// 9 6 3

File: Data Structures & Algorithms/permutations/submission-0.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    private void dfs(int[] nums, int begin, ArrayList<Integer> arr) {
        if(begin < 0) {
            res.add(Arrays.stream(nums).boxed().collect(Collectors.toCollection(ArrayList::new)));
            return;
        }
        for(int i = begin; i>= 0; i--) {
            swap(nums, i, begin);
            dfs(nums, begin - 1, arr);
            swap(nums, i, begin);
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        // res = new ArrayList<>();
        if(nums.length == 0) {
            return res;
        }
        dfs(nums, nums.length - 1, new ArrayList<Integer>());
        return res;
    }
}

// 1 2 3
// 2 1 3
// 1 2 3
// 1 3 2
//
